### ISSUE TITLE
xtcp polling performance

### ISSUE THREAD

**Issue by ccrome** [xmos/lib_xtcp#15]
Hi there,
  with the latest and greatest code, we get okay performance with lib_xtcp.  However, the throughput leaves a bit to be desired.

I believe this is in part due to the length of the polling loop in xtcp().  Clearly, it's greatly improved from the previous revision, which is good, but it could still use a little help.

To that end, added a timer in the polling loop to see where the time is spent, and the results are:

Op | Time for this op (uS) | Cumulative time (uS)
----|----------------------------|--------------------------
xtcp_service_clients|.75|.75
xtcp_check_connection_poll|3.96|4.71
uip_xtcp_checkstate|.35|5.06
xtcp_process_udp_acks|.81|5.87
start of next loop (i.e. time to  drop through the select  statement)|.26|6.13

At 100MBit, an average packet of 1000 byts takes 80 uS on the wire.  with a polling interval of 6 uS, this is maybe not a terrible ratio.

However, at 1000MBit, that same packet takes only 8uS, in which case the library when under load is spending a significant amount of time in xtcp_check_connection_poll. 

Obviously, for smaller packets, the problem is much worse.  (i.e. a 100 byte packet is only 0.8uS on the wire whereas the polling interval could be 10 times longer!)

Is there anything that can be done to improve performance here?  It would be great if we could make this thing event driven rather than polled.

BTW, I checked this both when idle, and when the ethernet is under load at with ping -f and the timings are very consistent.

**Comment by ccrome** [comment]
Just a quick note:  with my reflector app here (https://github.com/ccrome/xtcp-bug), the performance for gigabit is pretty much identically the same for 100MBit boards.  The max I get is about 1MBytes/second transferred in each direction.

This is just barely enough for our app, and doesn't leave room for our next one, which is going to be at least twice as big.

Thanks,
 -Caleb

**Comment by pthedinger** [comment]
Unfortunately, XTCP is a port on top of the UIP stack which is not designed to be event-driven. Therefore, making it fully event driven would require significant re-write. However, there might be a simpler option if what you are wanting to do is send and receive high bandwidth UDP traffic.

It might be possible to filter the UDP traffic at the level of the ethernet library so that it does not go through the XTCP stack, but instead goes directly to a user application. Can you see any reasons why that would not be a better approach?

**Comment by ccrome** [comment]
I was wondering about that.  I don't know networking that well, but I did a little experiement where I was running packets directly with the MII interface.  That's what you're suggesting right?

The thing I couldn't get to work is to get the TCP and ethernet interface both listening to the correct packet types.

I think it'll work otherwise.  

So, how do I get the TCP library working (so the webserver can serve up and respond to settings), and also have a second application that ethernet_tx and ethernet_rx interfaces listening for the UDP packets?

When I registered a macaddr with add_macaddr_filter() filter for my direct ethernet_rx interface, it slurped up all the packets, and TCP didn't work any more.  

Maybe with an ethertype filter?  What's an ethertype?  :-)    I'm not familiar with that term.  Back to google I guess.

Thanks,
 -Caleb

**Comment by pthedinger** [comment]
At the moment, the lib_ethernet layer only supports filtering of two types, MAC address or Ethertype (https://en.wikipedia.org/wiki/Ethernet_frame).

The Ethertype field for TCP / UDP packets have the same Ethertype of 0x0800 (https://en.wikipedia.org/wiki/EtherType). In order to differentiate UDP packets and only re-direct those to a second client will require deeper packet inspection. 

The UDP packets will be indicated by a value of 17 (0x11) in the protocol field of the IP header: https://en.wikipedia.org/wiki/IPv4#Header

There is always the complication that the header can be VLAN tagged, in which case an extra word is inserted in the packet before the Ethertype field with the Ethertype field having been changed to 0x8100. If you know that your frames are not going to be VLAN tagged then you can safely ignore this extra complication.

So, I think the best way to filter off the UDP packets will be to modify the ethernet_do_filtering() function in macaddr_filter.xc. The return value of the function is a word with 1-hot bit for each interface to which the packet must be sent. So, if you have two interfaces you'll be wanting to send UDP to interface 0x2 and all other packets for the mac address to interface 0x1 (depending on the order of interfaces).

You can optimise the function to only look for a single MAC address, and remove merge the Ethertype filtering from mii_ethernet_aux() function. So filtering is done once. The Ethertype extraction code is currently in mii_ethernet_aux().

Sorry I haven't had time to test this myself yet, and we have holidays here tomorrow/Monday for Easter. I can look at it more next week if it isn't too late. Hope that points you in the right direction.

**Comment by ccrome** [comment]
Heh, I was just looking at ethernet_do_filtering() to see what had to change.  Thanks for the pointers.  I will see what I can do. 

Happy Easter!  
A year ago or so, on my drive home, a bear appeared one day where a redwood used to stand.  Yesterday I snapped this photo.

![bubba](https://cloud.githubusercontent.com/assets/286973/14025910/861b0550-f1ad-11e5-87d0-c9355b9d3b1d.png)

**Comment by Helitune-RobMcKay** [comment]
I have changed the xtcp.xc file to use a faster polling rate and things still seem to work, just a bit faster
[xtcp.xc.txt](https://github.com/xmos/lib_xtcp/files/375795/xtcp.xc.txt)

**Comment by ccrome** [comment]
Yeah, I've done that too.  However, a polled architecture is simply not suitable for streaming audio I think, which is what I was trying to do.  We had to switch to USB.

Thanks for looking.

-Caleb

**Comment by humphrey-xmos** [comment]
Closing issue as old. 
For reference `lib_xtcp` version 7 now only supports LwIP, testing shows it supports greater than 50 Mbps in both directions at the same time, on xcore.ai using RMII 100BASE-T MAC.
